documentation/src/main/asciidoc/topics/con_performance_data_consistency.adoc

```diff
@@ -0,0 +1,105 @@
+How much contention depends on whether {brandname} locks keys optimistically or pessimistically.
+
+* Optimistic locking
++
+For low rates of contention, {brandname} locks keys when transactions invoke the `commit()` method.
+Keys are locked for shorter periods of time, which means less overall latency for write operations, but transaction failures are more costly.
+
+* Pessimistic locking
++
+For high rates of contention, {brandname} locks keys when transactions invoke the `put()` method.
+Keys are locked for longer periods of time, which increases latency for write operations, but makes transactions more efficient.
+
+Because they take longer to complete than performing individual operations, transaction failures are more costly in terms of excessive resource consumption.
+For example if a transaction takes five milliseconds to complete, with optimistic locking, you lose the entire five milliseconds.
+With pessimistic locking, if {brandname} detects a conflict after two milliseconds and the transaction rolls back, you lose only those two milliseconds instead.
+
+From that perspective having more efficient transactions with pessimistic locking might offset the reduction in overall write throughput.
+
+.Read isolation
+
+Isolation levels guarantee whether or not data in the cache has changed during transactions.
+
+.Isolation levels
+[%autowidth,cols="1,1",stripes=even]
+|===
+|`REPEATABLE_READ`
+|Read operations can only ever have one value in the context of a transaction. Once it retrieves the value of an entry, {brandname} returns that same value for any subsequent read operations.
+
+|`READ_COMMITTED`
+|Read operations can return different values if another transaction modifies keys. Each read operation in a transaction returns the most recent value for entries.
+|===
+
+In terms of overall latency for reads, neither isolation level incurs a penalty.
+However {brandname} can lock keys on writes but not on reads, which is required for `REPEATABLE_READ` isolation.
+If {brandname} cannot lock keys on reads, it needs another way to guarantee that values have not changed.
+
+To achieve this with optimistic locking, {brandname} uses write skew detection.
+Write skews are flaws in data consistency that happen when separate transactions simultaneously read the same key and then modify its value.
+
+[source,java,options="nowrap",subs=attributes+]
+----
+transaction1.begin()                    transaction2.begin()
+// Transaction 1 and 2 read "k1" simultaneously
+transaction1.get(k1)                    transaction1.get(k1)
+// Both transactions write different values to "k1"
+transaction1.put(k1, v2)                transaction1.put(k1, v3)
+transaction1.commit()                   transaction2.commit()
+----
+
+When write conflicts occur for a transactional cache, one of the transactions must roll back.
+Usually {brandname} can detect the order in which transactions occur and then roll back the older transaction.
```

Comment:
Infinispan does not detect any "order". Basically, the last transaction to invoke commit() loses or both loses.